mono/metadata/threads.c

```diff
@@ -6011,291 +6056,230 @@ mono_set_thread_dump_dir (gchar* dir) {
-			g_assertf(prev_state == MONO_SUMMARY_EMPTY, "Summary memory was not in a clean state prior to entry", NULL);
-
-			if (summarizing_thread_state != MONO_SUMMARY_EXPECT) {
-				const char *name = thread_summary_state_to_str ((MonoSummaryState)summarizing_thread_state);
-				g_error ("Status after init wrong: %s\n", name);
-			}
+	for (int i=0; i < state->nthreads; i++) {
+		sigprocmask (SIG_UNBLOCK, &sigset, &old_sigset);
 
-			summarizing_thread = tid;
-			mono_threads_pthread_kill (info, SIGTERM);
+		if (i == current_idx)
+			continue;
 
-			// Number of rounds of 40 milliseconds seconds to give each dumping thread
-			int count = 400;
+	#ifdef HAVE_PTHREAD_KILL
+		pthread_kill (state->thread_array [i], SIGTERM);
 
-			while (old_num_summarized == num_threads_summarized && count > 0) {
-				if (thread->state & (ThreadState_Unstarted | ThreadState_Aborted | ThreadState_Stopped))
-					break;
+		if (!state->silent)
+			MOSTLY_ASYNC_SAFE_PRINTF("Pkilling 0x%zx from 0x%zx\n", state->thread_array [i], current);
+	#else
+		g_error ("pthread_kill () is not supported by this platform");
+	#endif
+	}
+}
 
-				usleep (10);
-				mono_memory_barrier ();
-				const char *name = thread_summary_state_to_str ((MonoSummaryState)summarizing_thread_state);
-				MOSTLY_ASYNC_SAFE_PRINTF("Waiting for signalled thread %zx to collect stacktrace. Status: %s\n. Thread state: 0x%x\n", tid, name, thread->state);
-				count--;
-			}
+static void
+summarizer_post_dump (SummarizerGlobalState *state, MonoThreadSummary *this_thread, int current_idx)
+{
+	mono_memory_barrier ();
+	mono_atomic_store_ptr ((volatile gpointer *)&state->all_threads [current_idx], this_thread);
+}
 
-			// Free the gchandle; we're done waiting on this one, no matter how the waiting went.
-			mono_gchandle_free (handle);
+static void
+summarizer_state_term (SummarizerGlobalState *state, gchar **out)
+{
+	// See the array writes
+	mono_memory_barrier ();
 
-			if (count == 0) {
-				// After timeout
-				// Thread may have been in lock or something, may have died
-				// Either way, didn't respond. Rather than failing, we have to
-				// skip it.
-				mono_memory_barrier ();
-				switch (summarizing_thread_state) {
-				case MONO_SUMMARY_MUTATE_SHARED:
-					g_error ("Timed out when writing json log!");
-					break;
+	mono_summarize_native_state_begin ();
+	for (int i=0; i < state->nthreads; i++) {
+		MonoThreadSummary *thread = state->all_threads [i];
+		if (!thread)
+			continue;
 
-				case MONO_SUMMARY_EMPTY: {
-					g_assert (summarizing_thread == (MonoNativeThreadId) NULL);
-					break;
+		mono_summarize_native_state_add_thread (thread, thread->ctx);
+	}
+	*out = mono_summarize_native_state_end ();
 
-				case MONO_SUMMARY_EXAMINE:
-					MOSTLY_ASYNC_SAFE_PRINTF("Timed out, thread did not finish dumping\n");
+	// Clean up all of our state
+	memset (state, 0, sizeof (*state));
+	mono_atomic_store_i32 ((volatile gint32 *)&state->summary_state, 0);
+	mono_memory_barrier ();
```

Comment:
I think `mono_memory_barrier ()` isn't needed after you use `mono_atomic_{store,load}*`.

Reply:
@luhenry and @jaykrell may know the answer here. In channel I've heard that it might not. I think the reasoning being that the synchronization primitive will synchronize that word in the cache but that it won't cause the setting thread to see modifications made by other threads in that time. 

Is there a good rule of thumb here about read and write barriers and atomic operations?

Reply:
Unclear.

https://stackoverflow.com/questions/2599238/are-memory-barriers-necessary-for-atomic-reference-counting-shared-immutable-dat

Says atomics are not barriers.

https://gcc.gnu.org/onlinedocs/gcc-4.4.3/gcc/Atomic-Builtins.html
"In most cases, these builtins are considered a full barrier"

https://docs.microsoft.com/en-us/windows/desktop/sync/synchronization-functions
Provides versions with full barriers, half barriers, and no barrier.
Though they aren't full barriers on Xbox360 PowerPC I believe.

Reply:
I'm thinking that we may want barriers after atomic operations then.

Reply:
@luhenry has informed me that our wrapped implementations are expected to have a memory barrier. 

https://github.com/mono/mono/issues/10651 has been filed around our one clear exception to this rule